agg-ast/schema_derivation/src/schema_derivation_tests/match_expr.rs

```diff
@@ -1675,3 +1675,1085 @@ mod misc_ops {
+                input = format!(
+                    r#"{{"$match": {{"$expr": {{"{}": ["$foo", "$bar"]}}}}}}"#,
+                    $op
+                )
+                .as_str(),
+                ref_schema = Schema::Any
+            );
+            test_derive_schema_for_match_stage!(
+                $name_str_eq_null,
+                expected = Ok(Schema::Document(Document {
+                    keys: map! {
+                        "foo".to_string() => STRING_OR_NULL.clone(),
+                    },
+                    required: set!(),
+                    ..Default::default()
+                })),
+                input = format!(
+                    r#"{{"$match": {{"$expr": {{"$eq": [null, {{"{}": ["$foo", "$bar"]}}]}}}}}}"#,
+                    $op
+                )
+                .as_str(),
+                ref_schema = Schema::Any
+            );
+        };
+    }
+    macro_rules! test_derive_schema_for_unary_string_expression_match {
+        ($name_str:ident, $name_str_eq_null:ident, $op:expr ) => {
+            test_derive_schema_for_match_stage!(
+                $name_str,
+                expected = Ok(Schema::Document(Document {
+                    keys: map! {
+                        "foo".to_string() => Schema::Atomic(Atomic::String),
+                    },
+                    required: set!("foo".to_string()),
+                    ..Default::default()
+                })),
+                input = format!(r#"{{"$match": {{"$expr": {{"{}": "$foo"}}}}}}"#, $op).as_str(),
+                ref_schema = Schema::Any
+            );
+            test_derive_schema_for_match_stage!(
+                $name_str_eq_null,
+                expected = Ok(Schema::Document(Document {
+                    keys: map! {
+                        // This looks strange, but these ops actually crash the pipeline if their
+                        // arguments are NULL, or return Integers even if their arguments are NULL.
+                        "foo".to_string() => Schema::Atomic(Atomic::String),
+                    },
+                    required: set!("foo".to_string()),
+                    ..Default::default()
+                })),
+                input = format!(
+                    r#"{{"$match": {{"$expr": {{"$eq": [null, {{"{}": "$foo"}}]}}}}}}"#,
+                    $op
+                )
+                .as_str(),
+                ref_schema = Schema::Any
+            );
+        };
+    }
+    use super::*;
+    test_derive_schema_for_binary_string_expression_match!(concat, concat_eq_null, "$concat");
+
+    test_derive_schema_for_match_stage!(
+        strcasecmp,
+        expected = Ok(Schema::Document(Document {
+            keys: map! {
+                "foo".to_string() => Schema::Atomic(Atomic::String),
+            },
+            required: set!("foo".to_string()),
+            ..Default::default()
+        })),
+        input = r#"{"$match": {"$expr": {"$strcasecmp": ["$foo", "$bar"]}}}"#,
```

Comment:
what happens to $bar here -- should it not be in the result set? maybe we can change this (and the `test_derive_schema_for_binary_string_expression_match`) to use the `starting_schema` argument instead and have foo be something like any, and have bar be ... something else, maybe an anyof of something?

Reply:
I was just writing the same comment in the macro above but stopped since I saw this comment. My understanding is that since `bar` is not in the initial schema at all, it is assumed to not be in the output schema. Although, if it is guaranteed to be missing I feel like that should have some impact on the schema derivation -- the expression should definitely resolve to `null` if one of the arguments is definitely `missing`, right?

In terms of the test, agreed that we should make a change. Either update to use `starting_schema` or use a constant value instead of a field ref in the expression.

Reply:
yeah, I guess I wasn't thinking or communicating super clearly in this comment either. Because this stage is of course valid, and the type of `$bar` is simply missing. Which is totally fine, but even setting it as `null` (not required) should function the same for most of our cases and be easier to read